### ISSUE TITLE
Resnet architecture question

### ISSUE THREAD

**Issue by bilipa** [matterport/Mask_RCNN#117]
I found your resnet units was conv-bn-relu structure
why dont you use bn-relu-conv?
Does the second one have some disadvantage？
Thanks

**Comment by philferriere** [comment]
ResNets use two types of blocks, **identity blocks** and **convolutional blocks**:
<img width="729" alt="idblock3_kiank" src="https://user-images.githubusercontent.com/6441756/33684878-9bb8a89c-da84-11e7-8057-fc75c40aa778.png">

<img width="647" alt="convblock_kiank" src="https://user-images.githubusercontent.com/6441756/33685068-3f7ed2da-da85-11e7-8ee9-98f13dff8556.png">

As far as I know, they always use CONV-BN-ReLU layers, as originally designed by Kaiming He et al. You can confirm that by looking at the model definition they used here:

https://github.com/KaimingHe/deep-residual-networks/blob/master/prototxt/ResNet-50-deploy.prototxt
https://github.com/KaimingHe/deep-residual-networks/blob/master/prototxt/ResNet-101-deploy.prototxt
https://github.com/KaimingHe/deep-residual-networks/blob/master/prototxt/ResNet-152-deploy.prototxt

**Comment by bilipa** [comment]
@philferriere 
Yes, but I see this https://github.com/TuSimple/mx-maskrcnn/blob/master/rcnn/symbol/symbol_mask_fpn.py
and this paper https://arxiv.org/pdf/1603.05027.pdf
Maybe this(bn->relu->conv) structure can get a better result ?

**Comment by philferriere** [comment]
Thanks for the links, @bilipa. This is quite interesting, actually. The choice of that alternate design for a residual unit seems to be motivated by this paper: [Identity Mappings in Deep Residual Networks](https://arxiv.org/abs/1603.05027), by Kaiming He et al when he working for Microsoft Research in 2016:

![figure1](https://user-images.githubusercontent.com/6441756/33695764-da9133e8-dab3-11e7-9765-10d2969213a3.png)

Per their analysis, the impact seems two-fold (easier training, better generalization). That benefit is mostly noticeable when using a very large number of layers:

![figure2](https://user-images.githubusercontent.com/6441756/33696349-bf540a44-dab6-11e7-9801-920b56cb6c41.png)

As they point out on page 13, "The original ResNet-152 [1] has top-1 error of 21.3% on a 320x320 crop, and our pre-activation counterpart has 21.1%. **The gain is not big on ResNet-152 because this model has not shown severe generalization difficulties**. However, the original ResNet-200 has an error rate of 21.8%, higher than the baseline ResNet-152. But we nd that the original ResNet-200 has lower training error than ResNet-152, suggesting that it suffers from overfitting."

Perhaps that is why for a number of layers below 200, most frameworks (such as Keras', Tensorflow's tf.slim and CNTK's) still use the "old" residual unit...

**Comment by bilipa** [comment]
@philferriere 
Yes, in Keras and Tensorflow, I couldn't find this structure and the pretrained weight.
In my opinion, the TuSimple use the second structure, it should have some reason
The paper said this structure can help training more easier. Because my GPU only have 8G, the result of ResNet-50-FPN I trained is not good. So I wonder this structure whether good for my result.
I found the rois may exclude some target because of the ROIS_PER_IMAGE = 128.
Does you have a better method to improve the result? Thanks.

**Comment by philferriere** [comment]
Honestly, the only reason I would look at switching to the other type of residual units would be if I **already had good** training results with the current architecture, to see if it could generalize even better. If I had bad training results, like you seem to be experiencing, switching to the alternate residual units is definitely not the first thing I would look at. Rather, I would investigate what is unique about my dataset that makes the training fail. Either way, I wish you good luck with your project, @bilipa!

**Comment by bilipa** [comment]
Thanks! I will try investigate my dataset

**Comment by waleedka** [comment]
@bilipa Good question! I actually spent a lot of time deciding on which Resnet version to use. I ended up using v1 for these reasons:
1. The v2 (i.e. pre-activation) architecture improves things significantly for the 1001-layer network, and only a little bit for the 200-layer network (according to the paper). Improvements for layers < 200 are not there or very small.
2. Keras uses v1 and provides Imagenet weights for it, so that makes things easier.
3. Although the Mask RCNN paper doesn't explicitly mention which version of Resnet they used, they do reference the original Resnet paper and they don't reference the v2 paper. So it's likely that they also used the v1 version.

And I agree with @philferriere comment that if you don't see good results it's likely not due to using v1 or v2. Check your training data, and check the model hyper parameters.
